### PR TITLE
Importing changes

### PR DESCRIPTION
**Pull request by Yehonal**
* implemented nested mutation for all kind of associations: 1:1 , 1:N , N:M
* implemented transaction: true argument that enable transaction on mutations
* implemented set: true argument to set associations (instead of add) during mutations
* implemented "exclude" and "only" for attributes
* fixed allowNull bug on Input types
* Implemented custom Scalar types
* n:m relation now are fully supported
* id in input type can be nullable when autoincrement
* some code optimization
* allow graphql null when db default field value exists and not null enabled
* improved README
* implemented exclude/only for create/fetch/update
* implemented inputUpdateTypes (because edit mutations can't have NotNull fields otherwise you're forced to update all fields)
* exclude createdAt and updatedAt from graphql, they are handled by sequelize internally
* more flexible custom types option (with legacy support)
* fixed update/delete with composite primary keys
* Implemented subscriptions
* Implemented bulk deletes
* tons of other fixes

### CODE REVIEW

**Comment by Yehonal** [comment]
Please @alirizwan help me to fix conflicts and adapt tests.

I've vastly tested all my changes on a production project and now you can do EVERYTHING, even most complex mutation.

**Comment by alirizwan** [comment]
when running this I am getting `Schema must contain unique named types but contains multiple types named "ProductstatesEnumType".`

File: src/sequelize-graphql-schema.js [hunk 1]
```diff
@@ -579,9 +870,9 @@ const generateModelTypes = (models, remoteTypes) => {
       const cache = {};
```
Comment (alirizwan) [inline src/sequelize-graphql-schema.js]: move `const cache = {}` to outside for loop `for (let modelName in models) {` to avoid enum caching bug..

**Comment by Yehonal** [comment]
> I am not sure if it's a good idea to override attributeFields and use a custom implementation. Can you please explain why do we need that? Also I don't see it having any extra code than the original one.

because of this: https://github.com/mickhansen/graphql-sequelize/issues/649 read the entire issue.

I've explained to the author that this implementation is to use graphql-sequelize with mutations, but they do not support mutations, only queries...instead our library does so it's needed to avoid unwanted behavior with updates mostly.

there's also an opened PR but they didn't accept it yet: https://github.com/mickhansen/graphql-sequelize/pull/650

File: src/sequelize-graphql-schema.js [hunk 2]
```diff
@@ -346,33 +567,50 @@ const generateTypesFromObject = function(remoteData){
...
-    if (!isInput) {
+    if (isInput) {
+      if (associationType === "BelongsToMany") {
+        const aModel = relation.through.model;
+        if (!aModel.graphql)
+          aModel.graphql=defaultModelGraphqlOptions;
+        // if n:m join table, we have to create the connection input type for it
+        const _name = getTypeName(aModel, isInput, false, true);
+        if (!types[_name]) {
+          var t = generateGraphQLType(aModel, types, cache, isInput, false, assocModel, model)
```
[the remark this answers is not in the document]
Comment (Yehonal) [reply src/sequelize-graphql-schema.js]: working on

File: src/sequelize-graphql-schema.js (hunk 1 above)
Comment (Yehonal) [reply src/sequelize-graphql-schema.js]: working on

**Comment by Yehonal** [comment]
> when running this I am getting `Schema must contain unique named types but contains multiple types named "ProductstatesEnumType".`

i'll test and try to fix it

**Comment by Yehonal** [comment]
> > when running this I am getting `Schema must contain unique named types but contains multiple types named "ProductstatesEnumType".`
> 
> i'll test and try to fix it

Can you tell me how to do the tests? which commands i've to run?

**Comment by alirizwan** [comment]
@Yehonal is it ready to go?

**Comment by Yehonal** [comment]
> @Yehonal is it ready to go?

Yes, of course we've to solve merging conflicts and fix test classes after.
However i'm using my fork in production for 3 huge projects with all kind of queries on very complex database and it works great for now :)

**Comment by alirizwan** [comment]
@Yehonal can you please help me resolve conflicts?

**Comment by Antiavanti** [comment]
@Yehonal @alirizwan Hi guys, any progress here? I'm waiting especially for those changes related to the transactions (using `Sequalize.useCLS` together with hooks is really hard).

**Comment by alirizwan** [comment]
> @Yehonal @alirizwan Hi guys, any progress here? I'm waiting especially for those changes related to the transactions (using `Sequalize.useCLS` together with hooks is really hard).

HiAntivanti, I will be working on this tomorrow on making code more standardized with easy automatic transactions. Since automatic transactions are a bit pain in the ass and it's more on sequelize part, it's kind of known issue that cls don't work with async/await so making it easy to pass manual transactions. Control over transactions is important and expect a big update this week.

**Comment by Antiavanti** [comment]
@alirizwan Many thanks for a quick answer. Maybe at least switch to `cls-hooked` instead of `continuation-local-storage`? It's just one drop in replacement and async/await hooks will work like a charm. It might be of course temporary solution but at least it will work. I don't want to fork or publish own version just with changed dependency (`cls-hooked` is suggested even by Sequalize http://docs.sequelizejs.com/manual/transactions.html).

**Comment by Yehonal** [comment]
Could you please set the review as resolved and merge the PR?

**Pull request merged**